packages/headless/src/controllers/generated-answer/headless-generated-answer.ts

```diff
@@ -0,0 +1,164 @@
+import {EventSourcePolyfill} from 'event-source-polyfill';
+import {
+  GeneratedAnswerCitation,
+  GeneratedAnswerCitationsPayload,
+  GeneratedAnswerMessagePayload,
+} from '../../api/generated-answer/generated-answer-event-payload';
+import {SearchEngine} from '../../app/search-engine/search-engine';
+import {
+  updateMessage,
+  updateError,
+  streamAnswer,
+  resetAnswer,
+  SSEErrorPayload,
+  updateCitations,
+  setIsLoading,
+} from '../../features/generated-answer/generated-answer-actions';
+import {
+  logDislikeGeneratedAnswer,
+  logLikeGeneratedAnswer,
+  logOpenGeneratedAnswerSource,
+  logRetryGeneratedAnswer,
+} from '../../features/generated-answer/generated-answer-analytics-actions';
+import {generatedAnswerReducer as generatedAnswer} from '../../features/generated-answer/generated-answer-slice';
+import {GeneratedAnswerState} from '../../features/generated-answer/generated-answer-state';
+import {executeSearch} from '../../features/search/search-actions';
+import {GeneratedAnswerSection} from '../../state/state-sections';
+import {loadReducerError} from '../../utils/errors';
+import {buildController} from '../controller/headless-controller';
+
+export type {GeneratedAnswerState, GeneratedAnswerCitation};
+
+/**
+ * @internal
+ */
+export interface GeneratedAnswer {
+  /**
+   * The state of the GeneratedAnswer controller.
+   */
+  state: GeneratedAnswerState;
+  /**
+   * Re-executes the last query to generate an answer.
+   */
+  retry(): void;
+  /**
+   * Logs a custom event indicating a generated answer is relevant.
+   */
+  likeGeneratedAnswer(): void;
+  /**
+   * Logs a custom event indicating a generated answer is irrelevant.
+   */
+  dislikeGeneratedAnswer(): void;
+  /**
+   * Logs a custom event indicating a cited source link was clicked.
+   */
+  logCitationClick(citation: GeneratedAnswerCitation): void;
+}
+
+/**
+ * @internal
+ */
+export function buildGeneratedAnswer(engine: SearchEngine): GeneratedAnswer {
+  if (!loadGeneratedAnswerReducer(engine)) {
+    throw loadReducerError;
+  }
+
+  const {dispatch} = engine;
+  const controller = buildController(engine);
+  const getState = () => engine.state;
+
+  let timeout: ReturnType<typeof setTimeout>;
+  let source: EventSourcePolyfill;
+  let lastRequestId: string;
+  let lastStreamId: string;
+
+  const onMessage = (payload: GeneratedAnswerMessagePayload) =>
+    dispatch(updateMessage(payload));
+
+  const onCitations = (payload: GeneratedAnswerCitationsPayload) =>
+    dispatch(updateCitations(payload));
+
+  const onError = (error: SSEErrorPayload) => {
```

Comment:
Most of what's happening here to check what's the status on the source, update state by dispatching actions, actually looks like logic that should be done in a reducer, and not a controller 🤔

Reply:
I agree though the source object is not serializable AFAIK so I couldn't put it in the state. So the controller is where I chose to keep its reference but if you can think of a better place I'd like to hear it.

Also the EventSource object needs to be given the event handlers which do the action dispatching. They're defined here mainly because the controller is the object that has both access to the source and the dispatch function. I could alternatively pass the dispatch function to the client when starting the streaming to that it can do some dispatching. Do you think that's better?

Reply:
Or could this be done in the asyncThunk (streamAnswer) ?
You can dispatch action from inside an asyncThunk.

Reply:
If i can share the reference to `source` between the asyncthunk and the controller this may indeed work and would clean up the controller. I'll give it a try

Reply:
I tried it and it is much better! thanks for the suggestion